inception-recommendation-api/src/main/resources/META-INF/asciidoc/developer-guide/implementing_recommender_tutorial.adoc

```diff
@@ -0,0 +1,362 @@
+_Deeplearning4j_.
+
+Recommenders in INCEpTION heavily rely on https://uima.apache.org[Apache UIMA] types and features.
+A recommender is configured for a certain layer and a certain feature. A layer can be seen as the
+type of annotation you want to to, e.g. `POS`, `NER`. Layers correspond to UIMA types. A feature is
+one piece of information that should be annotated, e.g. the POS tag. One layer can have many features.
+
+Annotations are given to a recommender in the form of a
+http://uima.apache.org/d/uimaj-current/apidocs/index.html[UIMA CAS]. One CAS corresponds to one
+document in INCEpTION. Annotations from a CAS can be read and manipulated via the
+https://uima.apache.org/d/uimafit-current/api/org/apache/uima/fit/util/CasUtil.html[CasUtil].
+
+We start by creating a new class `DataMajorityNerRecommender` that implements `RecommendationEngine`.
+Please see the JavaDoc of the respective methods for their semantics.
+
+....
+public class DataMajorityNerRecommender
+    implements RecommendationEngine
+{
+    public static final Key<DataMajorityModel> KEY_MODEL = new Key<>("model");
+
+    private final Logger log = LoggerFactory.getLogger(getClass());
+
+    private final String layerName;
+    private final String featureName;
+
+    public DataMajorityNerRecommender(Recommender aRecommender)
+    {
+        layerName = aRecommender.getLayer().getName();
+        featureName = aRecommender.getFeature().getName();
+    }
+
+    void train(RecommenderContext aContext, List<CAS> aCasses) {}
+    void predict(RecommenderContext aContext, CAS aCas) {}
+    double evaluate(List<CAS> aCasses, DataSplitter aDataSplitter) { return 0.0; }
+
+    private static class DataMajorityModel {
+        private final String majorityLabel;
+        private final double confidence;
+
+        private DataMajorityModel(String aMajorityLabel, double aConfidence) {
+            majorityLabel = aMajorityLabel;
+            confidence = aConfidence;
+        }
+    }
+
+    private static class Annotation {
+        private final String label;
+        private final int begin;
+        private final int end;
+        private double score;
+
+        private Annotation(String aLabel, int aBegin, int aEnd)
+        {
+            label = aLabel;
+            begin = aBegin;
+            end = aEnd;
+        }
+    }
+}
+....
+
+For the constructor, we take the `Recommender` object which contains the recommender configuration,
+e.g. the layer and the name of the feature to recommend. The next step is to implement the required
+methods.
+
+`DataMajorityModel` and `Annotation` are internal data classes to simplify the code.
+
+=== RecommenderContext
+
+Instances of https://github.com/inception-project/inception/blob/master/inception-recommendation-api/src/main/java/de/tudarmstadt/ukp/inception/recommendation/api/recommender/RecommenderContext.java[RecommendationEngine] itself are stateless. If data like trained models need to be
```

Comment:
I'd avoid links to the code since these do not change automatically when we refactor.